packages/rest/src/request-context.ts

```diff
@@ -6,17 +6,90 @@
 import {Context} from '@loopback/context';
 import * as onFinished from 'on-finished';
 import {RestBindings} from './keys';
+import {RestServerResolvedConfig} from './rest.server';
 import {HandlerContext, Request, Response} from './types';
 
 /**
  * A per-request Context combining an IoC container with handler context
  * (request, response, etc.).
  */
 export class RequestContext extends Context implements HandlerContext {
+  /**
+   * Get the protocol used by the client to make the request.
+   * Please note this protocol may be different from what we are observing
+   * at HTTP/TCP level, because reverse proxies like nginx or sidecars like
+   * Envoy are switching between protocols.
+   */
+  get requestedProtocol(): string {
+    return (
+      (this.request.get('x-forwarded-proto') || '').split(',')[0] ||
+      this.request.protocol ||
+      this.serverConfig.protocol ||
+      'http'
+    );
+  }
+
+  /**
+   * Get the effective base path of the incoming request. This base path
+   * combines `baseUrl` provided by Express when LB4 handler is mounted on
+   * a non-root path, with the `basePath` value configured at LB4 side.
+   */
+  get basePath(): string {
+    const request = this.request;
+    let basePath = this.serverConfig.basePath || '';
+    if (request.baseUrl && request.baseUrl !== '/') {
+      basePath = request.baseUrl + basePath;
+    }
+    return basePath;
+  }
+
+  /**
+   * Get the base URL used by the client to make the request.
+   * This URL contains the protocol, hostname, port and base path.
+   * The path of the invoked route and query string is not included.
+   *
+   * Please note these values may be different from what we are observing
+   * at HTTP/TCP level, because reverse proxies like nginx are rewriting them.
+   */
+  get requestedBaseUrl(): string {
```

Comment:
Shouldn't we add unit test for this? We want to be able to detect any unexpected change in future.